Peer cache

pwojcikdev opened the pull request:
This introduces a new `peer_cache` component that is responsible for managing cached peers. Previously storing peers was done by `tcp_channels` and loading by `node`, which is not the core responsibility of these classes.
Previous behavior was that only the latest set of peers was stored (ie. cache was cleared before every write). Now the cache keeps track of peers up to a certain cutoff age (by default 1 hour).
Another behavior change is that network reachout thread tries to contact cached peers. Previously this was only done once, on node startup. This should help with resiliency if for some reason node temporarily loses connection to its peers (eg. due to network outage).

pwojcikdev committed:
Reach out to cached peers

pwojcikdev committed:
Test reconnecting to cached peers

pwojcikdev committed:
Headers for forward declarations

pwojcikdev committed:
Conversions for endpoint_key

pwojcikdev committed:
Time helpers

pwojcikdev committed:
Store timestamps in peer store

pwojcikdev committed:
Introduce `peer_cache`

pwojcikdev committed:
Tests for peer cache

dsiganos commented on nano/store/lmdb/peer.cpp:
```diff
@@ -4,12 +4,25 @@
 nano::store::lmdb::peer::peer (nano::store::lmdb::component & store) :
 	store{ store } {};
```
spurious semicolon
lesser point: the braces placement looks a bit strange stylistically

dsiganos commented:
Should `nano_node --debug_peers` also print the timestamp?
The current print line of --debug_peers can be simplified to this:
```
std::cout << i->first.endpoint () << std::endl;
```
@gr0vity-dev  heads-up: this might affect your scripts, I think you interact directly with the peers table, right?

gr0vity-dev commented:
Thanks for the heads-up.
The continuous testing scripts don't use the peers table directly anymore.

clemahieu reviewed (dismissed):
This looks like a good cleanup.

Only thing I noticed was on naming, is this really a cache? Maybe peer_history?

Not a blocker.

dsiganos commented on nano/node/network.cpp:
```diff
@@ -214,6 +215,40 @@ void nano::network::run_reachout ()
+		}
+		lock.unlock ();
+
+		node.stats.inc (nano::stat::type::network, nano::stat::detail::loop_reachout);
```
Should this counter be loop_reachout_cached?

dsiganos commented:
We do not really need to store the timestamp into the database, it can be kept in RAM only.
Is it stored in the database for design simplicity?

dsiganos commented on nano/node/peer_cache.cpp:
```diff
@@ -0,0 +1,156 @@
+	}
+}
+
+nano::error nano::peer_cache_config::serialize (nano::tomlconfig & toml) const
```
Are these not implemented on purpose? As a job for later?

pwojcikdev replied:
I added TODO comments to implement the config serialization.

dsiganos commented:
I think we should be storing the peer node ID in the peer database too.
But that is possibly beyond the scope of this PR.

pwojcikdev committed:
Fixes

pwojcikdev committed:
Rename to `peer_history`

pwojcikdev committed:
Merge branch 'develop' into cached-peers-2

# Conflicts:
#	nano/lib/thread_roles.cpp
#	nano/lib/thread_roles.hpp

pwojcikdev commented:
I fixed a few minor issues pointed by the review and renamed class to `peer_history`. The timestamps are stored in database for simplicity, storing them in ram could work but I don't think it's that important to optimize this. In the grand scheme of things storing an additional uint64 per peer is not going to make a difference.
